src/ck-core/cklocation.h

```diff
@@ -301,25 +311,50 @@ enum CkElementCreation_t : uint8_t
 class CkLocCache : public CBase_CkLocCache
 {
 private:
+  // This is the entry in the location table, which stores the PE and epoch number of the
+  // latest location update.
+  // TODO: If we can template CkLocCache on this type, we could store more useful
+  // information for specific entities being managed. ie: for arrays it could store
+  // indices as well. Could also store homePe if useful? Then one lookup would get you
+  // everything you need, rather than looking up in a bunch of tables.
+  // TODO: Is int the right type for pe and epoch?
+
   // Map of ID to PE
-  using IdPeMap = std::unordered_map<CmiUInt8, int>;
-  IdPeMap id2pe;
+  using LocationMap = std::unordered_map<CmiUInt8, CkLocEntry>;
+  LocationMap locMap;
 
   using Listener = std::function<void(CmiUInt8, int)>;
   std::list<Listener> listeners;
 
+  const CkLocEntry nullEntry = CkLocEntry();
 public:
   CkLocCache() = default;
   CkLocCache(CkMigrateMessage* m) : CBase_CkLocCache(m) {}
   ~CkLocCache() = default;
   void pup(PUP::er& p);
 
-  void inform(CmiUInt8 id, int nowOnPe);
-  int homePe(const CmiUInt8 id) const { return id >> CMK_OBJID_ELEMENT_BITS; }
   void requestLocation(CmiUInt8 id);
+
+  // Entry methods for updating location tables across PEs
   void requestLocation(CmiUInt8 id, int peToTell);
-  void updateLocation(CmiUInt8 id, int nowOnPe);
-  void erase(CmiUInt8 id) { id2pe.erase(id); }
+  void updateLocation(const CkLocEntry& e);
+
+  // Update the location table when an element migrates away
+  void recordEmigration(CmiUInt8 id, int pe);
+
+  // Query the local location table
+  const CkLocEntry& getLocationEntry(CmiUInt8 id) const
+  {
+    LocationMap::const_iterator itr = locMap.find(id);
+    return itr != locMap.end() ? itr->second : nullEntry;
+  }
+  int getPe(const CmiUInt8 id) const { return getLocationEntry(id).pe; }
+  int getEpoch(const CmiUInt8 id) const { return getLocationEntry(id).epoch; }
+  int homePe(const CmiUInt8 id) const { return id >> CMK_OBJID_ELEMENT_BITS; }
```

Comment:
Agreed. This should be further cleaned up in the 64 bit update, but for now I'll use the ObjID method. Especially because it further shields the code from 64bit ID constants/implementation.